packages/caravan-health/src/privacy.ts

```diff
@@ -0,0 +1,333 @@
+*/
+function getSpendTypeScore(
+  spendType: SpendType,
+  numberOfInputs: number,
+  numberOfOutputs: number,
+): number {
+  switch (spendType) {
+    case SpendType.PerfectSpend:
+      return 1 / 2;
+    case SpendType.SimpleSpend:
+      return 4 / 9;
+    case SpendType.UTXOFragmentation:
+      return 2 / 3 - 1 / numberOfOutputs;
+    case SpendType.Consolidation:
+      return 1 / numberOfInputs;
+    case SpendType.MixingOrCoinJoin:
+      let x = Math.pow(numberOfOutputs, 2) / numberOfInputs;
+      return ((2 / 3) * x) / (1 + x);
+    default:
+      throw new Error("Invalid spend type");
+  }
+}
+
+/*
+The transaction topology score evaluates privacy metrics based on the number of inputs and outputs.
+
+Expected Range : [0, 0.75]
+-> Very Poor : [0, 0.15]
+-> Poor : (0.15, 0.3]
+-> Moderate : (0.3, 0.45]
+-> Good : (0.45, 0.6]
+-> Very Good : (0.6, 0.75] 
+*/
+export async function getMeanTopologyScore(
+  transactions: Transaction[],
+  client: BlockchainClient,
+): Promise<number> {
+  let privacyScore = 0;
+  for (let tx of transactions) {
+    let topologyScore = await getTopologyScore(tx, client);
+    privacyScore += topologyScore;
+  }
+  return privacyScore / transactions.length;
+}
+
+export async function getTopologyScore(
+  transaction: Transaction,
+  client: BlockchainClient,
+): Promise<number> {
+  const numberOfInputs: number = transaction.vin.length;
+  const numberOfOutputs: number = transaction.vout.length;
+
+  const spendType: SpendType = determineSpendType(
+    numberOfInputs,
+    numberOfOutputs,
+  );
+  const score: number = getSpendTypeScore(
+    spendType,
+    numberOfInputs,
+    numberOfOutputs,
+  );
+
+  if (spendType === SpendType.Consolidation) {
+    return score;
+  }
+  for (let output of transaction.vout) {
+    let address = output.scriptPubkeyAddress;
+    let isResued = await isReusedAddress(address, client);
+    if (isResued === true) {
+      return score;
+    }
+  }
+  return score * DENIABILITY_FACTOR;
+}
+
+/* 
+In order to score for address reuse we can check the amount being held by reused addresses 
+with respect to the total amount
+
+Expected Range : [0,1]
+-> Very Poor : (0.8, 1]
+-> Poor : [0.6, 0.8)
+-> Moderate : [0.4, 0.6)
+-> Good : [0.2, 0.4)
+-> Very Good : [0 ,0.2) 
+*/
+export async function addressReuseFactor(
+  utxos: AddressUtxos,
+  client: BlockchainClient,
+): Promise<number> {
+  let reusedAmount: number = 0;
+  let totalAmount: number = 0;
+
+  for (const address in utxos) {
+    const addressUtxos = utxos[address];
+    for (const utxo of addressUtxos) {
+      totalAmount += utxo.value;
+      let isReused = await isReusedAddress(address, client);
+      if (isReused) {
+        reusedAmount += utxo.value;
+      }
+    }
+  }
+  return reusedAmount / totalAmount;
+}
+
+async function isReusedAddress(
+  address: string,
+  client: BlockchainClient,
+): Promise<boolean> {
+  let txs: Transaction[] = await client.getAddressTransactions(address);
```

Comment:
So, theoretically, if this function took an `address` and a list of `transactions`, it could check if each address was reused without making any network calls. This would eliminate any need for a `client` which simplifies the API quite a bit and makes it easier to use this library in other environments. Since the main function already requires basically the whole wallet transaction history, this isn't actually a huge deal since we should have that information already. 

The biggest downsides I could see are it's more complex to implement since you're not just checking how many receive txs the address got. Instead you've got to go through each txout, check the address version of the scriptpubkey and track how many it's received. Another challenge with this is that it's a somewhat expensive process (maybe not much more than making a network call though, but ideally you don't re-run for each address). 

A few ways to implement this I could imagine:

1. have a function that creates a map of addresses to booleans representing whether it's been reused or not. You loop through all txs in the wallet history and for each output, if the address is not in the map add it and set to false, if it is in the map then set it to true (alternatively the value can be an array of txids and if the list length is greater than 1 it is reused). This happens once and then this `isReusedAddress` function would simply take an address and this `reusedAddressMap`, check `reusedAddressMap.has(address)`, and if it was reused (value is true, or length is > 1).
2. You could do the above all in one function and rebuild the map each time you check for an address. Might be fine overall and actually less work for one off use cases, i.e. export the function for reuse factor to use on views for individual address scores like here:
<img width="1256" alt="Screenshot 2024-08-08 at 9 29 41 PM" src="https://github.com/user-attachments/assets/b5ef017f-1aaf-42d3-bfa6-005027e5d788">

3. You could have `addressReuseFactor` take a callback instead of `client` and then you can let it support multiple ways to determine if the address was reused. This one feels like it'd be a little messy though and overly complex

4. `getWalletPrivacyScore` already requires a list of transactions and addressUtxos as inputs. You could add something like the addressReuseMap as another input. You could also export a type of utility function as described in (1) as well as keep what you have here that uses the client and the consumer can build the map on their own if they want beforehand. 

5. OR instead of a third argument, you could reuse the `AddressUtxos` and maybe have it include txid info as well, something like an `AddressUsageMap` and it would be typed something like:

```typescript
interface UsageDetails {
  utxos: UTXO[];
  txids: string[];
}

interface AddressUsageMap {
  [address:string]: UsageDetails
}

// or instead of the second interface if you wanted to use an actual es6 Map:

type AddressUsageMap = Map<string, UsageDetails>;
const addressUsageMap: AddressUsageMap = new Map();

addressUsageMap.set("address1", { utxos: ["tx1"], txids: ["tx1"] });
```

6. It may be a little late now, but this could be an argument for making this all a single class that _could_ be instantiated with an optional client and use that (since with a class you don't have to pass the client down through several functions) OR us another of the methods above depending. Then you could instantiate the class once and call different scoring methods as desired. 

```typescript
const privacyCalculator = new PrivacyScore({...})
privacyCalculator.getPScore() 
privacyCalculator.getAddressReuseFactor()
privacyCalculator.getUtxoSpreadFactor()
// could return a map already passed in, use a client that was provided to calculate, or create the map fresh based on the first strategy and cache it for later use
privacyCalculator.getReusedAddresses() 
// if this is run before the previous function it would already have a usage map cached!
privacyCalculator.isAddressReused(address)
``` 

Why a change like this might this be worthwhile:

- It should simplify testing
- simplifies the API (currently a consumer outside of caravan has to have caravan/clients installed and create an instance of it) 
- Decouples this library completely from any backend or network which means you can run the analysis without an internet connection! 
- Some nice code cleanup. I think you had to pass the `client` through 3 or 4 layers of functions before it's actually used. This makes the code much less readable 
- Depending which approach you take, the consumer of the lib can choose however they want to get this information (even using a version of the function(s) you already have). 

Personally I think my preference is either something like (1) where you have a function that goes through the list of txs and creates a map of address to txid list or (5) having a  `AddressUsageMap` that can be passed in. 

Turning this all into a class is very appealing too because it allows more flexibility (the class could build the usage map from the transactions list if not provided or it could be passed in or it could use a client if available). But is definitely a much bigger re-write! So while I think this might be the best solution, I also think what you have here already is quite good as it is, so I'll leave that decision up to you.